algorithm/cgal/cgal.cc

```diff
@@ -3930,6 +3929,169 @@ void print_polygon_with_holes(
   std::cout << " }" << std::endl;
 }
 
+template <typename P>
+bool emitPolygonsWithHoles(const Plane& plane, const std::vector<P>& polygons,
+                           emscripten::val& emit_polygon,
+                           emscripten::val& emit_point) {
+  bool emitted = false;
+  for (const P& polygon : polygons) {
+    const auto& outer = polygon.outer_boundary();
+    emit_polygon(false);
+    emitted = true;
+    for (auto edge = outer.edges_begin(); edge != outer.edges_end(); ++edge) {
+      if (edge->source() == edge->target()) {
+        // Skip zero length edges.
+        std::cout << "QQ/skip zero length edge" << std::endl;
+        continue;
+      }
+      auto p = plane.to_3d(Point_2(edge->source().x(), edge->source().y()));
+      auto p2 = plane.to_3d(Point_2(edge->target().x(), edge->target().y()));
+      if (p == p2) {
+        continue;
+      }
+      emitPoint(p, emit_point);
+    }
+    for (auto hole = polygon.holes_begin(); hole != polygon.holes_end();
+         ++hole) {
+      emit_polygon(true);
+      emitted = true;
+      for (auto edge = hole->edges_begin(); edge != hole->edges_end(); ++edge) {
+        if (edge->source() == edge->target()) {
+          // Skip zero length edges.
+          std::cout << "QQ/skip zero length edge" << std::endl;
+          continue;
+        }
+        auto p = plane.to_3d(Point_2(edge->source().x(), edge->source().y()));
+        emitPoint(p, emit_point);
+      }
+    }
+  }
+  return emitted;
+}
+
+template <class Curve, class Curve_point>
+void emitCircularCurve(Curve& curve, Curve_point& position,
+                       Polygon_2& linear_polygon) {
+  // General loss of precision.
+  const auto& source = curve->source();
+  const auto& target = curve->target();
+  const auto& circle = curve->supporting_circle();
+  double cx = CGAL::to_double(circle.center().x());
+  double cy = CGAL::to_double(circle.center().y());
+  double sx = CGAL::to_double(source.x());
+  double sy = CGAL::to_double(source.y());
+  double tx = CGAL::to_double(target.x());
+  double ty = CGAL::to_double(target.y());
+  double source_angle = atan2(sy - cy, sx - cx);
+  double target_angle = atan2(ty - cy, tx - cx);
+  std::vector<Point_2> plan;
+  if (circle.orientation() == CGAL::CLOCKWISE) {
+    plan.push_back(Point_2(sx, sy));
+    if (target_angle > source_angle) {
+      target_angle -= M_PI * 2;
+    }
+    const double step = M_PI / 32.0;
+    double ox = sx - cx;
+    double oy = sy - cy;
+    double angle_limit = source_angle - target_angle;
+    for (double angle = step; angle < angle_limit; angle += step) {
+      plan.push_back(Point_2(cos(-angle) * ox - sin(-angle) * oy + cx,
+                             sin(-angle) * ox + cos(-angle) * oy + cy));
+    }
+    plan.push_back(Point_2(tx, ty));
+  } else {
+    // COUNTER-CLOCKWISE
+    plan.push_back(Point_2(sx, sy));
+    if (target_angle < source_angle) {
+      target_angle += M_PI * 2;
+    }
+    const double step = M_PI / 32.0;
```

Comment:
Is there any value to this being proportional to the diameter? Probably not worth the extra complexity

Reply:
I was thinking about it, but let's see how it goes first.